Add SIP-20 (External Data and Events) (#125)
MetaMask/SIPs: SIPS/sip-20.md

---
sip: 20

> title: External data (WebSockets and other custom data or events)

Reviewer: Maybe something like this?

```suggestion
title: External data entrypoint
```

Author: Updated.

status: Draft
author: David Drazic (@david0xd)
created: 2023-12-15
---

## Abstract

> This SIP proposes changes to the Snap manifest and additional request handler that allows Snap developers to implement
> functionality that is invoked when external data events occur. This proposal outlines some of the details around this feature.
> 
> The purpose of the initial version of this SIP is to provide support for WebSockets.
> This SIP only proposes one-way communication.
> This SIP is made with extensibility in mind and can be updated to handle more external data sources (like blockchain
> events, etc.).

Reviewer: ```suggestion
This SIP proposes a new permission that enables Snaps to receive data from
external sources. In the initial version, this SIP proposes support for
WebSockets, but it can be extended to support other data sources like
blockchain events in the future.

The Snap can specify the external data source in the Snap manifest. The
client then connects to the external data source and sends the data to the Snap.
The Snap can then do whatever it wants with the data.

This initial version only supports one-way communication from the external
source to the Snap. The Snap can't send any data back to the external source.
```

Author: Updated.

## Motivation

> Snaps are currently isolated from outside events and rely on combination of user actions or CRON jobs and fetch
> endowments in order to receive extra information from independent remote sources distributed over the internet.
> 
> This proposal provides an improvement that enables one way communication between a Snap and an external data or event
> source.
> The idea is that the external event source sends arbitrary data to the specific controller with listener running
> inside a wallet. The controller then sends the data to the Snap which runs execution of this feature specific RPC
> request handler.

Reviewer: ```suggestion
Snaps are currently limited in their ability to receive data from external
sources: Either they have to rely on user actions or cron jobs to fetch data, so
they can't react to events in real time. Snaps also cannot use WebSocket
connections to receive data from external sources, and are limited to HTTP
requests.
```

Author: Updated.

## Specification

> Formal specifications are written in Typescript.

### Language

> The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be
> interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

Reviewer: ```suggestion
The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in
uppercase in this document are to be interpreted as described in
[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
```

Author: Updated.

### Snap Manifest

> This SIP specifies a permission named `endowment:external-data`.

Reviewer: ```suggestion
This SIP introduces a new permission named `endowment:external-data`.
```

Author: Updated.

The new field can be specified as follows in a `snap.manifest.json` file:

```json
{
  "initialPermissions": {
    "endowment:external-data": [
      {
        "type": "websocket",
        "url": "wss://example.com/binary/endpoint",
        "dataType": "binary"
      },
      {
        "type": "websocket",
        "url": "wss://example.com/text/endpoint",
        "dataType": "text"
      },
      {
        "type": "blockchain",
        "chainId": "eip155:1",
        "events": [
          "block"
        ]
      }
    ]
  }
}

> ```

Reviewer: This looks like an example. We should start with describing the format of the permission, which is probably easier to do as a TypeScript type, i.e.:

```ts
/**
 * A WebSocket connection as specified in the Snap manifest.
 *
 * @property type - The type of connection. Currently, only WebSockets are
 * supported.
 * @property url - The URL to connect to. This must be a valid WebSocket URL,
 * starting with `wss://`. URLs starting with `ws://` are not allowed.
 * @property dataType - The type of data expected from the WebSocket. If this
 * field is omitted, text is assumed.
 */
type WebSocketConnection = {
  type: 'websocket';
  url: string;
  dataType?: 'text' | 'binary';
};

/**
 * An external data connection as specified in the Snap manifest.
 *
 * Currently, only {@link WebSocketConnection} is supported.
 */
type ExternalDataConnection = WebSocketConnection;

/**
 * External data connections as specified in the Snap manifest.
 *
 * This is the value of the `endowment:external-data` field in the Snap
 * manifest.
 */
type ExternalData = ExternalDataConnection[];
```

Reviewer: Then you can add a new section `#### Example` with the example.

Author: Done!

### Permission caveats

> Each caveat object MUST have `type` property that SHOULD be `websocket` or `blockchain`.

Reviewer: We should limit this SIP to just `websocket`.

Author: Updated. Only websocket left.

If the `type` is `websocket`then the caveat MUST have `url` and `dataType` properties.
The `url` MUST be string.
The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
The `chainId` MUST be string that represents supported blockchain.
The `events` SHOULD be an array of strings which represent name of the events for subscription.

- `type` property represents the type of data source.
- `url` is the WebSocket URL.
- `dataType` is type of data that WebSocket returns.
- `chainId` is the blockchain ID.
- `events` is the array of events to subscribe.

The caveats MUST NOT have duplicate objects with the same `url` or `chainId` properties.

The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.

> ### Snap implementation (request handler)

Reviewer: ```suggestion
### Snap implementation
```

Author: Updated.

> This SIP proposes the following RPC request handler to execute implemented functionality for external data events:

Reviewer: Again, I think we should start with a description of the new thing.

```suggestion
This SIP introduces a new `onExternalData` function export that MAY be
implemented by the Snap. The function is called every time the WebSocket client
receives some data. The function SHOULD accept an object parameter that MUST
have `data`, `type` and `source` properties.

The parameters represent the following:

- `data`: An object that represents the data that was received. Depending on
  the `data.type`, this SHOULD contain a `message` property, which SHOULD be
  either a `string` or an `ArrayBuffer`.
- `type`: The type of the external data source. For this SIP, this SHOULD always
  be `websocket`.
- `source`: The URL that the message was sent from. Snaps can use this to
  differentiate between different endpoints. This SHOULD be the exact same URL
  as specified in manifest.

The specification of types of `onExternalData` and its parameters:

```typescript
/**
 * A text message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string.
 */
type WebSocketTextMessage = {
  type: 'text';
  message: string;
};

/**
 * A binary message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 */
type WebSocketBinaryMessage = {
  type: 'binary';
  message: ArrayBuffer;
};

/**
 * A message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string or an ArrayBuffer, depending on
 * the type.
 */
type WebSocketData = {
  type: 'websocket';
  data: WebSocketTextMessage | WebSocketBinaryMessage;
};

/**
 * The data received from an external source.
 *
 * @property type - The type of the external data source.
 * @property data - The data received from the external data source.
 */
type ExternalData = WebSocketData;

type OnExternalDataHandlerArgs = ExternalData & {
  source: string;
};

type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
```

Author: Updated following your suggestion.

#### OnExternalDataHandler

Example:
```typescript
import { OnExternalDataHandler } from '@metamask/snaps-types';
import { assert } from '@metamask/utils';

export const onExternalData: OnExternalDataHandler = ({ id, type, data, source }) => {
  assert(type === 'websocket'); // `data` is inferred as `WebSocketData`.
  assert(data.type === 'text'); // `message` is inferred as `string`.

  // Now the Snap can do whatever is needed with the message.
  const json = JSON.parse(data.message);
};

> ```

Reviewer: I would move this to an `#### Example` section again.

Author: Updated.

Snap that wants to use external data sources MUST have `onExternalData` function export.

Exported function SHOULD accept an object parameter that MAY have `id`, `type`, `data` and `source` properties.

This export SHOULD be called every time the WebSocket client or Blockchain event listener receives some data or event.

The parameters represent the following:
- `id`: Unique ID of an event generated internally on the controller side. This can be used for tracking each event
  in case of storage needs or debugging. The `id` MUST be unique string.
- `data`: SHOULD be the raw message that the WebSocket client received. Depending on the `type`, this SHOULD be
  either a `string` or an `ArrayBuffer`.
- `type`: The type of the message that was sent. This SHOULD be either `text` or `binary`.
- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints.
  This SHOULD be the exact same URL as specified in manifest.

> The RECOMMENDED specification of types of `OnExternalDataHandler` and its parameters:

Reviewer: I don't think types should be a recommendation?

Author: Removed.

### Snap manifest

This SIP introduces a new permission named `endowment:external-data`.

```typescript
/**
 * A WebSocket connection as specified in the Snap manifest.
 *
 * @property type - The type of connection. Currently, only WebSockets are
 * supported.
 * @property url - The URL to connect to. This must be a valid WebSocket URL,
 * starting with `wss://`. URLs starting with `ws://` are not allowed.
 * @property dataType - The type of data expected from the WebSocket. If this
 * field is omitted, text is assumed.
 */
type WebSocketConnection = {
  type: 'websocket';
  url: string;
  dataType?: 'text' | 'binary';
};

/**
 * An external data connection as specified in the Snap manifest.
 *
 * Currently, only {@link WebSocketConnection} is supported.
 */
type ExternalDataConnection = WebSocketConnection;

/**
 * External data connections as specified in the Snap manifest.
 *
 * This is the value of the `endowment:external-data` field in the Snap
 * manifest.
 */
type ExternalData = ExternalDataConnection[];
```

#### Example

The new field can be specified as follows in a `snap.manifest.json` file:

> ```json

Reviewer 2: We may want to describe this as an object to more easily add more keys to the caveat.

For example, if we want to support SIP-21 here.

```json
{
  "initialPermissions": {
    "endowment:external-data": {
      "maxRequestTime": 50000,
      "connections": [
        {
          "type": "websocket",
          "url": "wss://example.com/binary/endpoint",
          "dataType": "binary"
        },
        {
          "type": "websocket",
          "url": "wss://example.com/text/endpoint",
          "dataType": "text"
        }
      ]
    }
  }
}
```

Author: Updated as you suggested.

Reviewer: Should probably be added to the definition then?

Author: Added as you suggested and answered here: https://github.com/MetaMask/SIPs/pull/125#discussion_r1494704575

 * @property message - The message as a string.
 */
type WebSocketTextMessage = {
  type: 'text';
  message: string;
};

/**
 * A binary message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 */
type WebSocketBinaryMessage = {
  type: 'binary';
  message: ArrayBuffer;
};

/**
 * A message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string or an ArrayBuffer, depending on
 * the type.
 */
type WebSocketData = {
  type: 'websocket';
  data: WebSocketTextMessage | WebSocketBinaryMessage;
};

/**
 * The data received from an external source.
 *
 * @property type - The type of the external data source.
 * @property data - The data received from the external data source.
 */
type ExternalData = WebSocketData;

type OnExternalDataHandlerArgs = ExternalData & {
  source: string;
};

type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
```

#### Example

```typescript
import { OnExternalDataHandler } from '@metamask/snaps-types';
import { assert } from '@metamask/utils';

> export const onExternalData: OnExternalDataHandler = ({ id, type, data, source }) => {

Reviewer 2: Where does `id` come from here? It isn't defined in the above types

Author: Probably left from some of the previous iterations. Removed now, seems to be redundant.

---
sip: 20
title: External data entry point
status: Draft
author: David Drazic (@david0xd)
created: 2023-12-15
---

## Abstract

This SIP proposes a new permission that enables Snaps to receive data from external sources.

In the initial version, this SIP proposes support for WebSockets, but it can be extended to support other data sources like blockchain events in the future.
The Snap can specify the external data source in the Snap manifest. The client then connects to the external data source and sends the data to the Snap.
The Snap can then do whatever it wants with the data. This initial version only supports one-way communication from the external source to the Snap. The Snap can't send any data back to the external source.

## Motivation

> Snaps are currently limited in their ability to receive data from external sources: Either they have to rely on user actions or cron jobs to fetch data, so they can't react to events in real time. Snaps also cannot use WebSocket connections to receive data from external sources, and are limited to HTTP requests.

Reviewer 3: ```suggestion
Snaps are currently limited in their ability to receive data from external sources; they have to rely on user actions or cron jobs to fetch data, so they can't react to events in real time. Snaps also cannot use WebSocket connections to receive data from external sources, and are limited to HTTP requests.
```

Author: Updated.

## Specification

> Formal specifications are written in Typescript.

### Language

The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

> ### Snap Manifest

Reviewer: ```suggestion
### Snap manifest
```

Author: Updated.

 * supported.
 * @property url - The URL to connect to. This must be a valid WebSocket URL,
 * starting with `wss://`. URLs starting with `ws://` are not allowed.
 * @property dataType - The type of data expected from the WebSocket. If this
 * field is omitted, text is assumed.
 */
type WebSocketConnection = {
  type: 'websocket';
  url: string;
  dataType?: 'text' | 'binary';
};

/**
 * An external data connection as specified in the Snap manifest.
 *
 * Currently, only {@link WebSocketConnection} is supported.
 */
type ExternalDataConnection = WebSocketConnection;

/**
 * External data connections as specified in the Snap manifest.
 *
 * This is the value of the `endowment:external-data` field in the Snap
 * manifest.
 */
type ExternalData = ExternalDataConnection[];
```

#### Example

The new field can be specified as follows in a `snap.manifest.json` file:

```json
{
  "initialPermissions": {
    "endowment:external-data": [
      {
        "type": "websocket",
        "url": "wss://example.com/binary/endpoint",
        "dataType": "binary"
      },
      {
        "type": "websocket",
        "url": "wss://example.com/text/endpoint",
        "dataType": "text"

>       },
>       {
>         "type": "blockchain",
>         "chainId": "eip155:1",
>         "events": [
>           "block"
>         ]
>       }

Reviewer: ```suggestion
      }
```

Author: Removed.

The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.

### Snap implementation

This SIP introduces a new `onExternalData` function export that MAY be implemented by the Snap. The function is called every time the WebSocket client receives some data. The function SHOULD accept an object parameter that MUST have `data`, `type` and `source` properties.
The parameters represent the following:
- `data`: An object that represents the data that was received. Depending on the `data.type`, this SHOULD contain a `message` property, which SHOULD be either a `string` or an `ArrayBuffer`.
- `type`: The type of the external data source. For this SIP, this SHOULD always be `websocket`.
- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints. This SHOULD be the exact same URL as specified in manifest.

The specification of types of `onExternalData` and its parameters:

```typescript
/**
 * A text message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string.
 */
type WebSocketTextMessage = {
  type: 'text';
  message: string;

> };

Reviewer: ```suggestion
};

```

Author: Updated.

/**
 * A binary message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 */
type WebSocketBinaryMessage = {
  type: 'binary';
  message: ArrayBuffer;

> };

Reviewer: ```suggestion
};

```

Author: Updated.

/**
 * A message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string or an ArrayBuffer, depending on
 * the type.
 */
type WebSocketData = {
  type: 'websocket';
  data: WebSocketTextMessage | WebSocketBinaryMessage;

> };

Reviewer: ```suggestion
};

```

Author: Updated.

/**
 * The data received from an external source.
 *
 * @property type - The type of the external data source.
 * @property data - The data received from the external data source.
 */

> type ExternalData = WebSocketData;

Reviewer: ```suggestion
type ExternalData = WebSocketData;

```

Author: Updated.

type OnExternalDataHandlerArgs = ExternalData & {
  source: string;

> };

Reviewer: ```suggestion
};

```

Author: Updated.

### Language

The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

### Snap manifest

This SIP introduces a new permission named `endowment:external-data`.

```typescript
/**
 * A WebSocket connection as specified in the Snap manifest.
 *
 * @property type - The type of connection. Currently, only WebSockets are
 * supported.
 * @property url - The URL to connect to. This must be a valid WebSocket URL,
 * starting with `wss://`. URLs starting with `ws://` are not allowed.
 * @property dataType - The type of data expected from the WebSocket. If this
 * field is omitted, text is assumed.
 */
type WebSocketConnection = {
  type: 'websocket';
  url: string;
  dataType?: 'text' | 'binary';
};

/**
 * An external data connection as specified in the Snap manifest.
 *
 * Currently, only {@link WebSocketConnection} is supported.
 */
type ExternalDataConnection = WebSocketConnection;

/**
 * External data connections as specified in the Snap manifest.
 *
 * This is the value of the `endowment:external-data` field in the Snap
 * manifest.
 */

> type ExternalData = ExternalDataConnection[];

Reviewer 2: ```suggestion
type ExternalData = { connections: ExternalDataConnection[] };
```

Author: Fixed.

> /**
>  * External data connections as specified in the Snap manifest.
>  *
>  * This is the value of the `endowment:external-data` field in the Snap
>  * manifest.
>  */
> type ExternalData = { connections: ExternalDataConnection[] };

Reviewer: ```suggestion
/**
 * External data connections as specified in the Snap manifest.
 *
 * This is the value of the `endowment:external-data` field in the Snap
 * manifest.
 *
 * @property maxRequestTime - The maximum request time for the `onExternalData`
 * entry point, as described in SIP-21.
 * @property connections - The external data connections.
 */
type ExternalData = {
  maxRequestTime: number;
  connections: ExternalDataConnection[]
};
```

Author: Updated. Just made `maxRequestTime` as optional.
